eth_account/hdaccount/deterministic.py

```diff
@@ -0,0 +1,217 @@
+    """
+    As specified by RFC4231 - https://tools.ietf.org/html/rfc4231
+    """
+    return hmac.new(chain_code, data, hashlib.sha512).digest()
+
+
+def ec_point(pkey: bytes) -> bytes:
+    """
+    Compute `point(p)`, where `point` is ecdsa point multiplication
+
+    Note: Result is ecdsa public key serialized to compressed form
+    """
+    return keys.PrivateKey(HexBytes(pkey)).public_key.to_compressed_bytes()
+
+
+class SoftNode(int):
+    """
+    Soft node (unhardened), where value = index
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index)
+        obj.index = index
+        return obj
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
+        return self.to_bytes(4, byteorder="big")
+
+    def encode(self) -> str:
+        return str(self.index)
+
+    def __repr__(self):
+        return f"SoftNode({self.index})"
+
+
+class HardNode(int):
+    """
+    Hard node, where value = index + BIP32_HARDENED_CONSTANT
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index + 0x80000000)  # 2**31, BIP32 "Hardening constant"
+        obj.index = index
+        return obj
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
+        return self.to_bytes(4, byteorder="big")
+
+    def encode(self):
+        return str(self.index) + "'"
+
+    def __repr__(self):
+        return f"HardNode({self.index})"
+
+
+def decode_node(node: str) -> Union[SoftNode, HardNode]:
+    if len(node) < 1:
+        raise ValueError("Cannot use empty string")
+    if node[-1] in ("'", "H"):
+        return HardNode(int(node[:-1]))
+    else:
+        return SoftNode(int(node))
+
+
+def derive_child_key(
+    parent_key: bytes,
+    parent_chain_code: bytes,
+    node: Union[SoftNode, HardNode],
+) -> Tuple[bytes, bytes]:
+    """
+    From BIP32:
+
+    The function CKDpriv((k_par, c_par), i) → (k_i, c_i) computes a child extended
+    private key from the parent extended private key:
+
+    - Check whether i ≥ 231 (whether the child is a hardened key).
+      - If so (hardened child):
+        let I = HMAC-SHA512(Key = c_par, Data = 0x00 || ser_256(kpar) || ser_32(i)).
+        (Note: The 0x00 pads the private key to make it 33 bytes long.)
+      - If not (normal child):
+        let I = HMAC-SHA512(Key = c_par, Data = ser_P(point(k_par)) || ser_32(i)).
+    - Split I into two 32-byte sequences, I_L and I_R.
+    - The returned child key k_i is parse_256(I_L) + k_par (mod n).
+    - The returned chain code c_i is I_R.
+    - In case parse_256(I_L) ≥ n or k_i = 0, the resulting key is invalid,
+      and one should proceed with the next value for i.
+      (Note: this has probability lower than 1 in 2**127.)
+    """
+    assert len(parent_chain_code) == 32
```

Comment:
Instead of `assert`ing throughout this function, I think it would be good to raise more specific errors here with nicer error messages. What do you think?

Reply:
Anywhere you see an `assert` I just wanted to be extra careful to make sure the sizes were correct. They should never be hit in practice, but would cause subtle errors if someone modified the code incorrectly.